pywr/optimisation/__init__.py

```diff
@@ -0,0 +1,109 @@
+from ..core import Model
+import logging
+logger = logging.getLogger(__name__)
+
+
+def cache_variable_parameters(model):
+    variables = []
+    variable_map = [0, ]
+    for var in model.variables:
+        size = var.double_size + var.integer_size
+
+        if size <= 0:
+            raise ValueError('Variable parameter "{}" does not have a size > 0.'.format(var.name))
+
+        variable_map.append(variable_map[-1] + size)
+        variables.append(var)
+
+    return variables, variable_map
+
+
+def cache_constraints(model):
+    constraints = []
+    for r in model.constraints:
+        constraints.append(r)
+
+    return constraints
+
+
+def cache_objectives(model):
+    # This is done to make sure the order is fixed during optimisation.
+    objectives = []
+    for r in model.objectives:
+        objectives.append(r)
+    return objectives
+
+
+# Global variables for individual process to cache the model and some of its data.
+# TODO make this a registry using a unique id for each model. This way multiple
```

Comment:
TODO

Reply:
What unique ID should we use? I have been thinking of adding a UID property to metadata using `uuid`. Though this would perhaps change upon `Model.reset`. Otherwise it could be model's name?

Reply:
I like the idea of using a UUID (`uuid.uuid4()`) which is initalised in the `__init__` method of `BaseOptimisationWrapper`. The global variables then become dictionaries.